VoyageAI embeddings support

ggozad opened the pull request:
<!-- Thank you for contributing to Pydantic AI! -->

- Add VoyageAIEmbeddingModel for VoyageAI's embedding API
- Support all current VoyageAI models including domain-specific ones (code, finance, law)
- Add voyageai optional dependency group

<!-- Please add the issue number that should be closed when this PR is merged. -->
<!-- If there is no issue, please create one first, even for simple bug fixes. -->
- Closes #3855 

### Pre-Review Checklist

<!-- These checks need to be completed before a PR is reviewed, -->
<!-- but you can submit a draft early to indicate that the issue is being worked on. -->

- [X] Any **AI generated code** has been reviewed line-by-line by the human PR author, who stands by it.
- [X] No **breaking changes** in accordance with the [version policy](https://github.com/pydantic/pydantic-ai/blob/main/docs/version-policy.md).
- [X] **Linting and type checking** pass per `make format` and `make typecheck`.
- [X] **PR title** is fit for the release changelog.

### Pre-Merge Checklist

<!-- These checks need to be completed before a PR is merged, -->
<!-- but as PRs often change significantly during review, -->
<!-- it's OK for them to be incomplete when review is first requested. -->

- [X] New **tests** for any fix or new behavior, maintaining 100% coverage.
- [X] Updated **documentation** for new features and behaviors, including docstrings for API docs.

Kludex commented on pydantic_ai_slim/pydantic_ai/embeddings/__init__.py:
```diff
@@ -44,6 +44,13 @@
+        'voyageai:voyage-code-3',
+        'voyageai:voyage-finance-2',
+        'voyageai:voyage-law-2',
+        'voyageai:voyage-code-2',
```
Is there an API to know which models are available?

ggozad replied:
Not that I know of :( 
According to this [discussion](https://docs.voyageai.com/discuss/66d697827e0adb0018cc9d07) making an api call with a non-existing model name returns also the list of valid models in the error message, but that sounds very fragile.

DouweM replied:
The error message in the cassette below has a few more names, should we include them all?

DouweM commented on pydantic_ai_slim/pydantic_ai/embeddings/voyageai.py:
```diff
@@ -0,0 +1,184 @@
+    usage_data = {'total_tokens': total_tokens}
+    response_data = {'model': model, 'usage': usage_data}
+
+    return RequestUsage.extract(
```
We should only use this if the models have data in genai-prices. In this case, it's better to build a `RequestUsage` object manually. As you can see in the tests snapshots, this will actually fail to extract anything.

DouweM commented:
@ggozad The precedent for a local-model provider that doesn't actually do much is `OutlinesProvider`, which only exists because of the `model_profile` method, and so that `OutlinesModel` can have consistent `__init__` signature with other models. 

So even though it's not strictly necessary, I'd prefer to keep the `SentenceTransformersProvider`. I could imagine us adding `EmbeddingModelProfile` for example, which would then need `SentenceTransformersProvider.model_profile`.

DouweM commented on pydantic_ai_slim/pydantic_ai/embeddings/__init__.py:
```diff
@@ -70,14 +77,21 @@ def infer_embedding_model(
 
         model_kind = normalize_gateway_provider(model_kind)
 
+    # Handle models that don't need a provider first
```
If we revert this change, would we get test coverage again?

DouweM commented on pydantic_ai_slim/pydantic_ai/providers/voyageai.py:
```diff
@@ -0,0 +1,74 @@
+            assert api_key is None, 'Cannot provide both `voyageai_client` and `api_key`'
+            assert base_url is None, 'Cannot provide both `voyageai_client` and `base_url`'
+            assert max_retries == 0, 'Cannot provide both `voyageai_client` and `max_retries`'
+            assert timeout is None, 'Cannot provide both `voyageai_client` and `timeout`'
```
Unless "most users" will need them, I'd prefer to _not_ expose all the arguments on `AsyncClient` as arguments here: users can just pass their own `voyageai_client` if they want this level of control.

ggozad replied:
Done, left only `api_key` & `voyageai_client`.

ggozad commented on tests/test_embeddings.py:
```diff
@@ -318,6 +322,63 @@ async def test_embed_error(self, co_api_key: str):
+                embeddings=IsList(IsList(IsFloat(), length=1024), length=2),
+                inputs=['hello', 'world'],
+                input_type='document',
+                usage=RequestUsage(),
```
The voyageAI API returns total_tokens=0 for document embeddings. I am not sure why this is, most likely a bug on their end.

DouweM commented on pydantic_ai_slim/pydantic_ai/embeddings/voyageai.py:
```diff
@@ -0,0 +1,192 @@
+    Defaults to False. If True, inputs that are too long will be truncated.
+    """
+
+    voyageai_input_type: VoyageAIEmbedInputType
```
Hmm if it only supports `query` and `document` anyway I don't think a setting is warranted. If "direct embedding without prefix" is something users would want to do, I think we should make it `'none'` instead of `None` so that the difference is more clear between this field being omitted (and we should use the default input type implied by the embed_query/document method) and explicitly being set to none.

ggozad commented:
@DouweM thank you for the thorough review, I think I addressed all your comments, merged the google embeddings changes. I also added the series 4 embedding models that were released yesterday I think.
